### DISCUSSION TITLE
race condition when registering interfaces - "interface input not found"

### DISCUSSION THREAD

**Discussion by SeanDylanGoff** [directus/directus#11254] [Feature Requests]
what i found out so far:

app/src/main.ts calls registerInterfaces(app)

while registerInterfaces(app) is in progress (awaiting the import()'s) the following happens:

- hydrate() is called by onBeforeEach guard in app/src/router.ts
- hydrate calls setLanguage in app/src/lang/set-language.ts which translates interfacesRaw.value and stores the result in interfaces.value

HOWEVER: as registerInterfaces(app) is still in progress this call simply sets interfaces.value to an empty array. this then gets picked up by the interface fallback handling which replaces the actual interface with "input" - and then fails aswell because "input" is missing, too.

i also noticed another problem with probably similar root cause: sometimes the theme "sticks" to light even though the user chose dark. toggling the theme fixes it.

does anyone else have the same problem? a possible fix could be to delay hydration until main.js is fully executed.

### To Reproduce

i don't know how to reliably reproduce this issue - it comes and goes in my dev instance, never saw it on production.
Once it happens it happens reliably - even surviving browser cache deletes and vite restarts. Restarting the browser itself solves it - so i guess the reason it's sticky could be V8 cache.

**Comment by azrikahar** [comment]
That sounds odd, but can't say I've come across this before. 🤔 Though to be fair I don't always have custom interface(s) in my past setups.

Does it only happen to one specific custom interface? Or you've faced this with various different custom interfaces?

Also wondering has this happened in the past and you just so happen to report this now since you're free to do so, or it specifically started happening on `v9.4.2`?

> it comes and goes in my dev instance, never saw it on production.
> custom docker image for development

Would you be able to share the Dockerfile for the custom docker image? May somehow be relevant as to why you're only facing in your said dev instance but not production.

**Comment by SeanDylanGoff** [comment]
> Does it only happen to one specific custom interface? Or you've faced this with various different custom interfaces?
can't really tell as i'm mostly working on one specific interface and can't really force the problem to occur on purpose anyways.
 
> Also wondering has this happened in the past and you just so happen to report this now since you're free to do so, or it specifically started happening on `v9.4.2`?
same problem - a lot of variables changed since i last wrote an interface:
- first time using v9.4.2
- interface is quite the chonker (12 vue files)
- i haven't done a lot of work with this development setup either (the dockerfile/my changes)

my changes (including dockerfile): https://github.com/cynova/directus/commit/63875eeadee519ca54f034143a08e07d09288f8e
image:
https://github.com/cynova/directus/pkgs/container/directus/12578083?tag=dev-9.4.2


> Would you be able to share the Dockerfile for the custom docker image? May somehow be relevant as to why you're only facing in your said dev instance but not production.

however, i think the most probable cause is my physical setup: i'm developing on a remote instance --> more latency. i think this is what might trigger the race condition. 

I added console.log's before and after interface registration, on route setup and inside onBeforeEach hook. 
![image](https://user-images.githubusercontent.com/6235675/148220317-cc625519-7688-4e6b-9bc4-44e8ad019fd0.png)

it's obvious that vue-router does not wait for init().



Possible fix:
i moved app.use(router); after registerInterfaces, panels etc. directly before app.mount.
This seems to solve the problem, and so far i haven't seen any downsides. i'm not sure if that's a clean solution though...

**Comment by rijkvanzanten** [comment]
@azrikahar I think this might be pretty similar to #10707

**Comment by SeanDylanGoff** [comment]
update: i continued working on the interface for long enough that i would expect the problem to occur and it didn't - so moving use(router) really seems to fix it for me. i'll report should this assumption turn out to be false.

**Comment by rijkvanzanten** [comment]
@SeanDylanGoff It makes sense to me to register all the extensions before we tell Vue to fire up, though we'll have to do a bit more testing to see if that has any other unintended side effects (cc @nickrum)

**Comment by nickrum** [comment]
We can't fully register all extensions before initializing Vue though, because most extensions need to globally register components.

**Comment by rijkvanzanten** [comment]
> We can't fully register all extensions before initializing Vue though, because most extensions need to globally register components.

Yeah! But in this case, we'd initialize Vue first, then register the extensions, and then start the router 🤔

**Comment by nickrum** [comment]
I'd rather fix this in a different way. All the App initialization code feels quite unintuitive. But, I don't see why starting the router afterwards wouldn't work 👍

**Comment by rijkvanzanten** [comment]
I'll move this to a discussion so we can keep discussing the various improvements in the loading strategy to optimize for these (edge) cases 🙂

**Comment by claytongulick** [comment]
I'm running into the same issue:

<img width="1440" alt="Screen Shot 2022-07-11 at 11 35 46 AM" src="https://user-images.githubusercontent.com/695284/178313887-be7f0513-b7e4-468a-b3b2-6ab727588e6e.png">

We have lots of custom interfaces loading, there's an issue with the build that's not de-duplicating deps because of the way rollup is called internally, it creates a 14MB (!!) index.js file for the interfaces.

With that file downloading and loading so slowly, we hit the race condition on the interfaces and directus isn't usable.

**Reply by rijkvanzanten** [reply]
Interesting, we'll have to do some debugging what's causing the race condition to happen. The registration is a promise that's awaited before the app is mounted:

https://github.com/directus/directus/blob/033e12323e001cda3bf36c86e87da15398d72f20/app/src/main.ts#L47-L56

and the download of the interfaces bundle itself should similarly be awaited here:

https://github.com/directus/directus/blob/033e12323e001cda3bf36c86e87da15398d72f20/app/src/interfaces/register.ts#L16

🤔

**Reply by claytongulick** [reply]
Hmm. It's definitely related to registering the custom interfaces, as soon as I rename the extensions folder and restart directus, the issue disappears.

**Reply by claytongulick** [reply]
I'll dig into it today, I'm in all that code right now trying to figure out how to simplify and optimize the build and deploy process.

**Reply by claytongulick** [reply]
I think one issue is here:
In extensions.ts (controller):
```
                 const extensionManager = getExtensionManager();

		const extensionSource = extensionManager.getAppExtensions(type);
		if (extensionSource === undefined) {
			throw new RouteNotFoundException(req.path);
		}
```

This calls getAppExtensions synchronously, however the extension manager loads the extensions asynchronously on startup, and especially with a change and lengthly dynamic internal rollup call, the extension source may not be available.

I'm not sure why the problem persists after startup though, some sort of caching issue?

**Reply by rijkvanzanten** [reply]
Ohhh interesting, so the problem isn't the way the app loads them in, it could be the way the API is providing it!
